src/analyzer.ts

```diff
@@ -77,37 +113,61 @@ export class Analyzer {
 
     // Note we can't use Promise.all here, because new promises can be added to
     // the promises array as long as any promise is pending.
-    const errors = new Set<Failure>();
-    while (this.#placeholderUpgradeFailureSinks.length > 0) {
-      const result = await this.#placeholderUpgradeFailureSinks.shift();
-      if (result?.ok === false) {
-        for (const error of result.error) {
-          errors.add(error);
-        }
-      }
-    }
-    if (errors.size > 0) {
-      for (const error of errors) {
-        const supercedes = (error as Partial<DependencyOnMissingPackageJson>)
-          .supercedes;
-        if (supercedes != null) {
-          errors.delete(supercedes);
-        }
+    await this.#waitForAnalysisToComplete();
+    {
+      const errors = await this.#getDiagnostics();
+      if (errors.size > 0) {
+        return {ok: false, error: [...errors]};
       }
-      return {ok: false, error: [...errors]};
     }
 
     // We can safely assume all placeholders have now been upgraded to full
     // configs.
-    const rootConfig = rootPlaceholder.placeholder as ScriptConfig;
+    const rootConfig = rootPlaceholder.placeholder;
+    if (rootConfig.state === 'unvalidated') {
+      throw new Error(
+        `Internal error: script ${root.name} in ${root.packageDir} is still unvalidated but had no failures`
+      );
+    }
     const cycleResult = this.#checkForCyclesAndSortDependencies(
       rootConfig,
       new Set()
     );
     if (!cycleResult.ok) {
       return {ok: false, error: [cycleResult.error]};
     }
-    return {ok: true, value: rootConfig};
+    const validRootConfig = cycleResult.value;
+    return {ok: true, value: validRootConfig};
+  }
+
+  async #getDiagnostics(): Promise<Set<Failure>> {
+    const failures = new Set<Failure>();
+    for await (const failure of this.#packageJsonReader.getFailures()) {
+      failures.add(failure);
+    }
+    for (const info of this.#placeholders.values()) {
+      for (const failure of info.placeholder.failures) {
+        failures.add(failure);
+      }
+    }
+    for (const failure of failures) {
+      const supercedes = (failure as Partial<DependencyOnMissingPackageJson>)
+        .supercedes;
+      if (supercedes != null) {
+        failures.delete(supercedes);
+      }
+    }
+    return failures;
+  }
+
+  async #waitForAnalysisToComplete() {
+    while (this.#ongoingWorkPromises.length > 0) {
+      const promise = this.#ongoingWorkPromises[0];
+      await promise;
+      if (promise === this.#ongoingWorkPromises[0]) {
```

Comment:
Need a comment here? Why wouldn't we just shift before awaiting?

Also should we pop, since the order shouldn't matter, and as you pointed out in another PR pop is faster?

Reply:
Yeah, this is tricky, and a bit trickier with pop(), but agree that's the better method to call.

Alternatively we might represent it as a Set instead of an array so we can quickly remove any element, but that doesn't have as nice of an API for "give me an arbitrary element" as an array has.

Reply:
Good explanation in comment ty